engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java

```diff
@@ -31,17 +32,30 @@ public void applyState(final long elementInstanceKey, final ProcessInstanceRecor
     elementInstanceState.newInstance(
         flowScopeInstance, elementInstanceKey, value, ProcessInstanceIntent.ELEMENT_ACTIVATING);
 
+    if (flowScopeInstance == null) {
+      // process instance level
+      return;
+    }
+
     final var flowScopeElementType = flowScopeInstance.getValue().getBpmnElementType();
     final var currentElementType = value.getBpmnElementType();
-    if (isContainerElement(flowScopeElementType, currentElementType)) {
-      // we currently spawn new tokens only for container elements
-      // we might spawn tokens for other bpmn element types as well later, then we can improve here
+
+    if (isContainerElement(flowScopeElementType, currentElementType)
+        || isNonInterruptingEventSubprocess(flowScopeInstance, currentElementType)
+        || currentElementType == BpmnElementType.BOUNDARY_EVENT
+        || currentElementType == BpmnElementType.INTERMEDIATE_CATCH_EVENT) {
+      // we currently spawn new tokens only for:
+      // container elements, non interrupting event sub process, boundary events and intermediate
+      // catch events
+      // we might spawn tokens for other bpmn element types as well later, then we can improve
+      // here
       elementInstanceState.spawnToken(flowScopeInstance.getKey());
```

Comment:
Maybe it's just me, but this spawn token stuff is really confusing imo, it's hard to keep track of :sweat_smile:

Reply:
Totally agree!!! @saig0 promised to me that we remove this after migration :crossed_fingers:

Reply:
Yes. I created an issue here: #6539
Let's do it after the migration :rocket: